src/JSInterop/Microsoft.JSInterop/src/DotNetObjectRefOfT.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Wraps a JS interop argument, indicating that the value should not be serialized as JSON
+    /// but instead should be passed as a reference.
+    ///
+    /// To avoid leaking memory, the reference must later be disposed by JS code or by .NET code.
+    /// </summary>
+    /// <typeparam name="TValue">The type of the value to wrap.</typeparam>
+    public sealed class DotNetObjectRef<TValue> : IDotNetObjectRef, IDisposable where TValue : class
+    {
+        private long? _trackingId;
+
+        /// <summary>
+        /// This A is for meant for JSON deserialization and should not be used by user code.
+        /// </summary>
+        [Obsolete("This API is meant for JSInterop infrastructure and should not be used by user code.")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public DotNetObjectRef()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DotNetObjectRef{TValue}" />.
+        /// </summary>
+        /// <param name="value">The value to pass by reference.</param>
+        public DotNetObjectRef(TValue value)
+        {
+            Value = value;
+            _trackingId = DotNetObjectRefManager.Current.TrackObject(this);
+        }
+
+        /// <summary>
+        /// Gets the object instance represented by this wrapper.
+        /// </summary>
+        [JsonIgnore]
+        public TValue Value
+        {
+            get;
+            // Workaround for https://github.com/dotnet/corefx/issues/37536 and https://github.com/dotnet/corefx/issues/37567
+            // Once fixed, we should make this property private.
```

Comment:
Could you make sure there's a bug tracking this, and that it's scheduled to be done before we ship 3.0? It's important that people don't actually write to `Value` as it completely weirdifies the programming model if they can.

Reply:
https://github.com/aspnet/AspNetCore/issues/10298